pattisdr/osf.io
# [Migration Prep Required] Feature/Registration Responses Validation [ENG-874]

❗️ Locks to the same release as ENG-842 -  this finishes adding being able to update registration_responses when updating drafts.  ENG-842 builds `registration_responses` on create, but not update. 

## Purpose

- Add validation of `registration_responses` field, which is the more flattened field storing responses to supplemental registration questions that corresponds to the more flattened schema blocks.
- Validation ensures that required fields are present (if `required_fields` is True), that multiple choice responses are one of the provided responses, and that file information is formatted properly.  Also doesn't allow any extra keys to go into the db, unless the key matches a `registration_response_key`

## Changes
- Builds jsonschema for validating `registration_responses` on schemas
- Validates `registration_responses` when a DraftRegistration is created or updated, and `registration_responses was passed in.  If old workflow (`registration_metadata`) was passed in, use existing validation.
- Validates that file information is correct on `registration_metadata`. Your filename, and sha256 have to be accurate to register.


## QA Notes
- This PR allows updating a draft to update either `registration_responses` or `registration_metadata` through the API.  If you update the new `registration_responses` field, it will be expanded and stored in `registration_metadata` (verify through the API).
- If you update the old Draft.registration_metadata field through the API, it should be flattened and likewise stored under `registration_responses`.
- Verify you can update both `registration_metadata`, and `registration_responses` on the Draft Registration serializer at the same time
- Verify that `registration_metadata` is no longer a required field when updating a draft. 
- `registration_responses` or `registration_metadata` on the DraftRegistration are validated when updating the draft, not both.  
- Verify that required fields are not required to be present when updating the draft
- Verify that you get 400 errors when trying to update `registration_responses` incorrectly. For example, your data is the wrong type, your answer choice doesn't match one of the multiple choice answers, or you include a question that doesn't exist.
- This PR is not validating when the registration is created (but this validation will be turned on in https://openscience.atlassian.net/browse/ENG-366)
-  Verify that invalid file information can't be sent in for a question. For example, the nodeId on the file has to be the target.  The `file_name` has to be the name of the file.  The sha256 has to correspond to a version of that file, otherwise, you cannot create a registration.

⭐️ This used to be caching the jsonschema on the RegistrationSchema, but that has been removed; we're calculating on the fly now.
## Documentation



## Side Effects



## Ticket

https://openscience.atlassian.net/browse/ENG-874
